blockchain/indexers/utreexocfindex.go

```diff
@@ -0,0 +1,301 @@
+	idx := dbTx.Metadata().Bucket(utreexoCFIndexParentBucketKey).Bucket(key)
+	return idx.Delete(h[:])
+}
+
+var _ Indexer = (*UtreexoCFIndex)(nil)
+
+var _ NeedsInputser = (*UtreexoCFIndex)(nil)
+
+type UtreexoCFIndex struct {
+	db          database.DB
+	chainParams *chaincfg.Params
+
+	chain *blockchain.BlockChain
+
+	utreexoProofIndex *UtreexoProofIndex
+
+	flatUtreexoProofIndex *FlatUtreexoProofIndex
+}
+
+func (idx *UtreexoCFIndex) NeedsInputs() bool {
+	return true
+}
+
+// Init initializes the utreexo cf index. This is part of the Indexer
+// interface.
+func (idx *UtreexoCFIndex) Init(_ *blockchain.BlockChain) error {
+	return nil // Nothing to do.
+}
+
+// Key returns the database key to use for the index as a byte slice. This is
+// part of the Indexer interface.
+func (idx *UtreexoCFIndex) Key() []byte {
+	return utreexoCFIndexParentBucketKey
+}
+
+// Name returns the human-readable name of the index. This is part of the
+// Indexer interface.
+func (idx *UtreexoCFIndex) Name() string {
+	return utreexoCFIndexName
+}
+
+// Create is invoked when the index manager determines the index needs to
+// be created for the first time. It creates buckets for the custom utreexo
+// filter index.
+func (idx *UtreexoCFIndex) Create(dbTx database.Tx) error {
+	meta := dbTx.Metadata()
+
+	utreexoCfIndexParentBucket, err := meta.CreateBucket(utreexoCFIndexParentBucketKey)
+	if err != nil {
+		return err
+	}
+
+	for _, bucketName := range utreexoCfHeaderKeys {
+		_, err = utreexoCfIndexParentBucket.CreateBucket(bucketName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// storeUtreexoCFilter stores a given utreexocfilter header
+func storeUtreexoCFHeader(dbTx database.Tx, block *btcutil.Block, filterData []byte,
+	filterType wire.FilterType) error {
+	if filterType != wire.UtreexoCFilter {
+		return errors.New("invalid filter type")
+	}
+
+	// Figure out which header bucket to use.
+	hkey := utreexoCfHeaderKeys[0]
+	h := block.Hash()
+
+	// fetch the previous block's filter header.
+	var prevHeader *chainhash.Hash
+	ph := &block.MsgBlock().Header.PrevBlock
+	if ph.IsEqual(&zeroHash) {
+		prevHeader = &zeroHash
+	} else {
+		pfh, err := dbFetchUtreexoCFilterIdxEntry(dbTx, hkey, ph)
+		if err != nil {
+			return err
+		}
+
+		// Construct the new block's filter header, and store it.
+		prevHeader, err = chainhash.NewHash(pfh)
+		if err != nil {
+			return err
+		}
+	}
+
+	fh, err := builder.MakeHeaderForUtreexoCFilter(filterData, *prevHeader)
+	if err != nil {
+		return err
+	}
+	return dbStoreUtreexoCFilterIdxEntry(dbTx, hkey, h, fh[:])
+}
+
+// ConnectBlock is invoked by the index manager when a new block has been
+// connected to the main chain.
+// This is part of the Indexer interface.
+func (idx *UtreexoCFIndex) ConnectBlock(dbTx database.Tx, block *btcutil.Block,
+	stxos []blockchain.SpentTxOut) error {
+
+	blockHash := block.Hash()
+	roots, leaves, err := idx.fetchUtreexoRoots(dbTx, blockHash)
```

Comment:
Wouldn't this fetch the wrong state at times? The code that calls the `ConnectBlock` for indexers is at https://github.com/utreexo/utreexod/blob/e166f1004e45c98c17d26ec7bad786519fd544a3/blockchain/chain.go#L747-L755

and it's in the order of the indexers appended here: https://github.com/utreexo/utreexod/blob/e166f1004e45c98c17d26ec7bad786519fd544a3/server.go#L3200-L3264

We want the utreexo roots after this block has been applied to the utreexo state. But for this we might be off by one and fetch the previous state.